### PR TITLE
URL Refactoring

### LINKED ISSUE #501: Improve URL Parser performance.
Optimise URL.fromString function
Write jmh benchmarks and try micro optimisations

### PR DESCRIPTION
**Pull request by ShrutiVerma97**
fixes #501 #846 
Benchmark results
```
[info] Benchmark                                Mode  Cnt     Score     Error  Units
[info] URLParserBenchmark.benchmarkURLParser   thrpt   10  1795.675 ? 142.343  ops/s
[info] URLParserBenchmark.benchmarkURLParser2  thrpt   10  2131.620 ?  71.715  ops/s
[info] URLParserBenchmark.benchmarkURLParser3  thrpt   10  1930.228 ?  83.573  ops/s
```

### CODE REVIEW

File: zio-http/src/main/scala/zhttp/http/Scheme.scala [hunk 1]
```diff
@@ -26,4 +26,11 @@ object Scheme       {
       case "HTTP"  => Option(HTTP)
       case _       => None
     }
+
+  def fromString2(scheme: String): Scheme =
+    scheme.toUpperCase match {
+      case "HTTPS" => HTTPS
+      case "HTTP"  => HTTP
+      case _       => null
```
Comment (amitksingh1490) [inline zio-http/src/main/scala/zhttp/http/Scheme.scala]: Try adding a subtype empty.

File: zio-http/src/main/scala/zhttp/http/URL.scala [hunk 2]
```diff
@@ -67,6 +67,30 @@ object URL {
...
+  private def fromAbsoluteURI2(uri: URI): URL = {
+
+    def portFromScheme(scheme: Scheme): Int = scheme match {
+      case Scheme.HTTP  => 80
+      case Scheme.HTTPS => 443
+      case null         => -1
+    }
+
+    val scheme = Scheme.fromString2(uri.getScheme)
+    val port   = if (uri.getPort != -1) uri.getPort else portFromScheme(scheme)
```
Comment (amitksingh1490) [inline zio-http/src/main/scala/zhttp/http/URL.scala]: try moving uri.getPort to val

Comment (amitksingh1490) [inline zio-http/src/main/scala/zhttp/http/URL.scala]: This code is also not safe since portFromScheme can alo return -1

Comment (ShrutiVerma97) [reply zio-http/src/main/scala/zhttp/http/URL.scala]: We are handling that in the next statement

**Comment by tusharmath** [comment]
One thing we can do @ShrutiVerma97 here is to convert URL to a sum type. 

```scala
sealed trait URL

final case class FromString(string: String) extends URL
final case class Cons(
  path: Path,
  kind: URL.Location = URL.Location.Relative,
  queryParams: Map[String, List[String]] = Map.empty,
  fragment: Option[Fragment] = None,
) extends URL
```

So this way we will keep an unparsed version as string, and parse it only when its truly needed.

**Comment by ShrutiVerma97** [comment]
Benchark results URL vs URL2 
```
[info] Benchmark                                Mode  Cnt     Score     Error  Units
[info] URLParserBenchmark.benchmarkURL2Parser  thrpt    5  2122.770 ? 302.008  ops/s
[info] URLParserBenchmark.benchmarkURLParser   thrpt    5  2033.823 ?  51.423  ops/s
[info] URLParserBenchmark.benchmarkURL2Parser2  thrpt    5  2315.495 ?  53.017  ops/s
[info] URLParserBenchmark.benchmarkURL2Parser3  thrpt    5  2060.433 ?  68.606  ops/s
```

File: zio-http/src/main/scala/zhttp/http/PathModule.scala [hunk 3]
```diff
@@ -63,8 +63,8 @@ private[zhttp] trait PathModule { module =>
   }
 
   object Path {
-    def apply(): Path                   = End
-    def apply(string: String): Path     = if (string.trim.isEmpty) End else Path(string.split("/").toList)
+    def apply(): Path               = End
+    def apply(string: String): Path = if (string.trim.isEmpty || string == null) End else Path(string.split("/").toList)
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/PathModule.scala]: this change should not be required

File: zio-http/src/main/scala/zhttp/http/Scheme.scala (hunk 1 above)
Comment (tusharmath) [reply zio-http/src/main/scala/zhttp/http/Scheme.scala]: Can be optimized further.

File: zio-http/src/main/scala/zhttp/http/URL5.scala [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    }
+  }
+
+}
+object URL5 {
+
+  def apply(path: Path): URL5                                = Cons(path = path)
+  def apply(str: String): Either[HttpError.BadRequest, URL5] = Decode(string = str).decode
+
+  final case class Decode(string: String) extends URL5 {
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/URL5.scala]:
```suggestion
  final case class Raw(string: String) extends URL5 {
```

**Comment by ShrutiVerma97** [comment]
Updated Benchmarks:
```
[info] Benchmark                                  Mode  Cnt    Score     Error  Units
[info] URLParserBenchmark.benchmarkURLOldParser  thrpt    3  423.353 ? 414.729  ops/s
[info] URLParserBenchmark.benchmarkURLParser     thrpt    3  456.072 ?  95.470  ops/s`

**Pull request closed without merging**